src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs

```diff
@@ -426,6 +427,92 @@ public async Task SslStream_ClientCertificate_SendsChain()
+
+            var serverOptions = new SslServerAuthenticationOptions();
+            serverOptions.ServerCertificateContext = SslStreamCertificateContext.Create(Configuration.Certificates.GetServerCertificate(), null);
+
+            (Stream clientStream, Stream serverStream) = TestHelper.GetConnectedTcpStreams();
+            using (clientStream)
+            using (serverStream)
+            using (SslStream client = new SslStream(new RandomIOStream(clientStream, readBufferSize)))
+            using (SslStream server = new SslStream(new RandomIOStream(serverStream)))
+            {
+                Task t1 = client.AuthenticateAsClientAsync(clientOptions, CancellationToken.None);
+                Task t2 = server.AuthenticateAsServerAsync(serverOptions, CancellationToken.None);
+
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(t1, t2);
+
+                Task writer = Task.Run(() =>
+                {
+                    Memory<byte> data = new Memory<byte>(dataToCopy);
+                    while (data.Length > 0)
+                    {
+                        int writeLength = Math.Min(data.Length, writeBufferSize);
+                        if (useAsync)
+                        {
+                            server.WriteAsync(data.Slice(0, writeLength)).GetAwaiter().GetResult();
+                        }
+                        else
+                        {
+                            server.Write(data.Span.Slice(0, writeLength));
+                        }
+
+                        data = data.Slice(Math.Min(writeBufferSize, data.Length));
+                    }
+
+                    server.ShutdownAsync().GetAwaiter().GetResult();
+                });
+
+                Task reader = Task.Run(() =>
+                {
+                    SHA256 hash = SHA256.Create();
+                    byte[] readBuffer = new byte[readBufferSize];
+                    int totalLength = 0;
+                    int readLength;
+
+                    while (true)
+                    {
+                        if (useAsync)
+                        {
+                            readLength = client.ReadAsync(readBuffer).GetAwaiter().GetResult();
+                        }
+                        else
+                        {
+                            readLength = client.Read(readBuffer);
+                        }
+
+                        if (readLength == 0)
+                        {
+                            hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+                            break;
+                        }
+
+                        hash.TransformBlock(readBuffer, 0, readLength, null, 0);
+                        totalLength += readLength;
+                        Assert.True(totalLength <= bufferSize);
+                    }
+
+                    Assert.Equal(bufferSize, totalLength);
+                    AssertExtensions.SequenceEqual(srcHash, hash.Hash);
```

Comment:
My personal opinion is that it's actually easier to just buffer the entire result and compare the entire result to the source data. This avoids dealing with hashes at all. And doing a bytewise comparison of a 16M buffer isn't *that* expensive.

But if you prefer the hash approach, I'm fine with that too.

Reply:
updated. I was thinking about it and the buffer companion may give clue where the difference is. The buffer size is somewhat arbitrary so we can trim it down if needed.